include/fiction/io/read_gate_level_layout.hpp

```diff
@@ -0,0 +1,771 @@
+                         signal             = signal->NextSiblingElement("signal"))
+                    {
+                        tile<Lyt> incoming_signal{};
+
+                        // get x-coordinate of incoming signal
+                        auto* const incoming_signal_x = signal->FirstChildElement("x");
+                        if (incoming_signal_x != nullptr and incoming_signal_x->GetText())
+                        {
+                            incoming_signal.x = std::stoull(incoming_signal_x->GetText());
+                        }
+                        else
+                        {
+                            throw gate_level_parsing_error("Error parsing gate_level file: no element 'x' in 'signal'");
+                        }
+
+                        // get y-coordinate of incoming signal
+                        auto* const incoming_signal_y = signal->FirstChildElement("y");
+                        if (incoming_signal_y != nullptr and incoming_signal_y->GetText())
+                        {
+                            incoming_signal.y = std::stoull(incoming_signal_y->GetText());
+                        }
+                        else
+                        {
+                            throw gate_level_parsing_error("Error parsing gate_level file: no element 'y' in 'signal'");
+                        }
+
+                        // get z-coordinate of incoming signal
+                        auto* const incoming_signal_z = signal->FirstChildElement("z");
+                        if (incoming_signal_z != nullptr and incoming_signal_z->GetText())
+                        {
+                            incoming_signal.z = std::stoull(incoming_signal_z->GetText());
+                        }
+                        else
+                        {
+                            throw gate_level_parsing_error("Error parsing gate_level file: no element 'z' in 'signal'");
+                        }
+
+                        gate.incoming.push_back(incoming_signal);
+                    }
+                }
+
+                gates.push_back(gate);
+            }
+
+            // sort gates ascending based on id
+            std::sort(gates.begin(), gates.end(), gate_storage::compare_by_id);
+
+            for (const auto& gate : gates)
+            {
+                const tile<Lyt> location{gate.loc.x, gate.loc.y, gate.loc.z};
+
+                if (gate.incoming.size() == 0)
+                {
+                    if (gate.type == "PI")
+                    {
+                        lyt.create_pi(gate.name, location);
+                    }
+                    else
+                    {
+                        throw gate_level_parsing_error("Error parsing gate_level file: unknown gate type: " +
+                                                       std::string(gate.type));
+                    }
+                }
+
+                else if (gate.incoming.size() == 1)
```

Comment:
Generally, for all these input sizes: what if both a gate name and a hex string are defined? And what if they don't match?

Reply:
The gate name or hex string are saved in the same field, therefore they can't be defined at the same time.

Reply:
Understood.